File: lib/logic/lyrics_parser.dart
import 'dart:convert';

/// Represents a single line of lyrics with optional timing.
class LyricsLine {
  final int? timeMs; // Time in milliseconds, null for plaintext
  final String text;

  LyricsLine({this.timeMs, required this.text});

  Map<String, dynamic> toJson() => {'time': timeMs, 'text': text};

  factory LyricsLine.fromJson(Map<String, dynamic> json) =>
      LyricsLine(timeMs: json['time'] as int?, text: json['text'] as String);
}

/// Represents parsed lyrics data.
class LyricsData {
  final String type; // 'timed' or 'plain'
  final List<LyricsLine> lines;

  LyricsData({required this.type, required this.lines});

  Map<String, dynamic> toJson() => {
    'type': type,
    'lines': lines.map((l) => l.toJson()).toList(),
  };

  factory LyricsData.fromJson(Map<String, dynamic> json) => LyricsData(
    type: json['type'] as String,
    lines: (json['lines'] as List)
        .map((l) => LyricsLine.fromJson(l as Map<String, dynamic>))
        .toList(),
  );

  String toJsonString() => jsonEncode(toJson());

  static LyricsData fromJsonString(String json) =>
      LyricsData.fromJson(jsonDecode(json) as Map<String, dynamic>);
}

/// Parser for various lyrics file formats.
class LyricsParser {
  /// Parse LRC format lyrics.
  /// Format: [mm:ss.xx] Lyrics text
  static LyricsData parseLrc(String content) {
    final lines = <LyricsLine>[];
    final regex = RegExp(r'\[(\d+):(\d+)\.?(\d+)?\](.*)');

    for (final line in content.split('\n')) {
      final match = regex.firstMatch(line.trim());
      if (match != null) {
        final minutes = int.parse(match.group(1)!);
        final seconds = int.parse(match.group(2)!);
        final centiseconds = int.tryParse(match.group(3) ?? '0') ?? 0;
        final text = match.group(4)?.trim() ?? '';

        // Convert to milliseconds
        final timeMs =
            (minutes * 60 * 1000) +
            (seconds * 1000) +
            (centiseconds * 10); // centiseconds to ms

        if (text.isNotEmpty) {
          lines.add(LyricsLine(timeMs: timeMs, text: text));
        }
      }
    }

    // Sort by time
    lines.sort((a, b) => (a.timeMs ?? 0).compareTo(b.timeMs ?? 0));

    return LyricsData(type: 'timed', lines: lines);
  }

  /// Parse SRT (SubRip) format.
  /// Format:
  /// 1
  /// 00:00:12,500 --> 00:00:15,000
  /// Lyrics text
  static LyricsData parseSrt(String content) {
    final lines = <LyricsLine>[];
    final blocks = content.split(RegExp(r'\n\s*\n'));
    final timeRegex = RegExp(
      r'(\d+):(\d+):(\d+)[,.](\d+)\s*-->\s*\d+:\d+:\d+[,.]?\d*',
    );

    for (final block in blocks) {
      final blockLines = block.trim().split('\n');
      if (blockLines.length >= 2) {
        // Find timestamp line
        for (int i = 0; i < blockLines.length; i++) {
          final match = timeRegex.firstMatch(blockLines[i]);
          if (match != null) {
            final hours = int.parse(match.group(1)!);
            final minutes = int.parse(match.group(2)!);
            final seconds = int.parse(match.group(3)!);
            final millis = int.parse(match.group(4)!.padRight(3, '0'));

            final timeMs =
                (hours * 3600 * 1000) +
                (minutes * 60 * 1000) +
                (seconds * 1000) +
                millis;

            // Text is everything after the timestamp line
            final text = blockLines.sublist(i + 1).join(' ').trim();
            if (text.isNotEmpty) {
              lines.add(LyricsLine(timeMs: timeMs, text: text));
            }
            break;
          }
        }
      }
    }

    // Sort by time
    lines.sort((a, b) => (a.timeMs ?? 0).compareTo(b.timeMs ?? 0));

    return LyricsData(type: 'timed', lines: lines);
  }

  /// Parse plaintext lyrics (no timing).
  static LyricsData parsePlaintext(String content) {
    final lines = content
        .split('\n')
        .map((l) => l.trim())
        .where((l) => l.isNotEmpty)
        .map((l) => LyricsLine(text: l))
        .toList();

    return LyricsData(type: 'plain', lines: lines);
  }

  /// Auto-detect format and parse.
  static LyricsData parse(String content, String filename) {
    final lowerFilename = filename.toLowerCase();

    if (lowerFilename.endsWith('.lrc')) {
      return parseLrc(content);
    } else if (lowerFilename.endsWith('.srt')) {
      return parseSrt(content);
    } else {
      // Check if content looks like LRC
      if (RegExp(r'\[\d+:\d+').hasMatch(content)) {
        return parseLrc(content);
      }
      // Check if content looks like SRT
      if (RegExp(r'\d+:\d+:\d+[,.]').hasMatch(content)) {
        return parseSrt(content);
      }
      // Default to plaintext
      return parsePlaintext(content);
    }
  }
}

File: logic/lyrics_parser.py
"""Lyrics Parser for GroovyBox.

This module handles parsing, serialization, and deserialization of lyrics
in multiple formats:
- LRC: Synchronized lyrics with timestamps (e.g., [01:23.45] lyrics text)
- SRT: Subtitle format with time ranges
- Plaintext: Unsynchronized plain text lyrics

Supports automatic format detection and JSON serialization for database storage.
"""

import re
import json
from data.models import LyricsLine, LyricsData


def parse_lrc(content: str) -> LyricsData:
    """Parse LRC format synchronized lyrics.
    
    Expected format: [MM:SS.xx] lyrics text
    
    Args:
        content: Raw LRC file content string.
    
    Returns:
        LyricsData with type="timed" and parsed lines.
    """
    lines = []
    regex = re.compile(r'\[(\d+):(\d+)\.?(\d+)?\](.*)')
    for line in content.split('\n'):
        m = regex.match(line.strip())
        if m:
            minutes = int(m.group(1))
            seconds = int(m.group(2))
            centiseconds = int(m.group(3) or '0')
            text = m.group(4).strip()
            # Convert to milliseconds
            time_ms = minutes * 60 * 1000 + seconds * 1000 + centiseconds * 10
            if text:
                lines.append(LyricsLine(time_ms=time_ms, text=text))
    lines.sort(key=lambda x: x.time_ms or 0)
    return LyricsData(type="timed", lines=lines)


def parse_srt(content: str) -> LyricsData:
    """Parse SRT format subtitle/lyrics.
    
    Expected format:
        1
        00:00:20,000 --> 00:00:24,400
        Lyrics text line 1
        Lyrics text line 2
    
    Args:
        content: Raw SRT file content string.
    
    Returns:
        LyricsData with type="timed" and parsed lines.
    """
    lines = []
    blocks = re.split(r'\n\s*\n', content)
    time_regex = re.compile(r'(\d+):(\d+):(\d+)[,.](\d+)\s*-->\s*\d+:\d+:\d+[,.]?\d*')
    for block in blocks:
        block_lines = block.strip().split('\n')
        for i, bl in enumerate(block_lines):
            m = time_regex.match(bl)
            if m:
                hours = int(m.group(1))
                minutes = int(m.group(2))
                seconds = int(m.group(3))
                millis = int(m.group(4).ljust(3, '0'))
                time_ms = hours * 3600 * 1000 + minutes * 60 * 1000 + seconds * 1000 + millis
                text = ' '.join(block_lines[i + 1:]).strip()
                if text:
                    lines.append(LyricsLine(time_ms=time_ms, text=text))
                break
    lines.sort(key=lambda x: x.time_ms or 0)
    return LyricsData(type="timed", lines=lines)


def parse_plaintext(content: str) -> LyricsData:
    """Parse plain text lyrics (unsynchronized).
    
    Each non-empty line becomes a lyrics line without timestamps.
    
    Args:
        content: Raw text content string.
    
    Returns:
        LyricsData with type="plain" and parsed lines.
    """
    lines = [
        LyricsLine(text=l.strip())
        for l in content.split('\n')
        if l.strip()
    ]
    return LyricsData(type="plain", lines=lines)


def parse(content: str, filename: str) -> LyricsData:
    """Parse lyrics content with automatic format detection.
    
    Uses the file extension to determine format, falling back to
    content-based detection if the extension is unrecognized.
    
    Args:
        content: Raw lyrics file content.
        filename: Original filename (used for extension detection).
    
    Returns:
        LyricsData with parsed lines and detected type.
    """
    lower = filename.lower()
    if lower.endswith('.lrc'):
        return parse_lrc(content)
    elif lower.endswith('.srt'):
        return parse_srt(content)
    else:
        # Content-based format detection
        if re.search(r'\[\d+:\d+', content):
            return parse_lrc(content)
        if re.search(r'\d+:\d+:\d+[,.]', content):
            return parse_srt(content)
        return parse_plaintext(content)


def lyrics_to_json(data: LyricsData) -> str:
    """Serialize LyricsData to a JSON string for database storage.
    
    Args:
        data: The LyricsData instance to serialize.
    
    Returns:
        JSON string representation of the lyrics.
    """
    return json.dumps({
        "type": data.type,
        "lines": [{"time": l.time_ms, "text": l.text} for l in data.lines],
    }, ensure_ascii=False)


def lyrics_from_json(s: str) -> LyricsData:
    """Deserialize LyricsData from a JSON string.
    
    Args:
        s: JSON string from the database.
    
    Returns:
        LyricsData instance reconstructed from the JSON.
    """
    d = json.loads(s)
    return LyricsData(
        type=d.get("type", "plain"),
        lines=[LyricsLine(time_ms=item.get("time"), text=item["text"]) for item in d.get("lines", [])],
    )
